### PR TITLE
Graduate indexed job to beta

### LINKED ISSUE #98434: Metrics for job-controller
**What would you like to be added**:

Metrics for the job controller operations, including:
- latency of a job sync `job_sync_duration_seconds`
- sync count `job_sync_total` (with a `result` label for `success`/`error`)
- TBD job finished count `job_finished_total`  (with a `result` label for `success`/`error`)

All metrics should include labels that allow to narrow delays or errors to features or code paths:
- `completion_mode`: `NonIndexed`, `Indexed`.
- TBD `pattern`: `single`, `workQueue`, `fixedCompletion`
- TBD type of event that triggered sync, and whether it was immediate.

**Why is this needed**:

This is essential to diagnose common usage patterns or problems in the controller.

**Note**:

A KEP will be written for this feature, but we want to gather feedback on other useful metrics or labels. The lists are currently open to additions or removals.

/sig apps
/area workload-api/job

### PR DESCRIPTION
**Pull request by AliceZhang2016**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Graduate indexed job from alpha to beta, contain three parts in total:
1. add three metrics to job controller:
a. job_sync_duration_seconds: tracks the latency of a Job sync.
b. job_sync_total: tracks the number of Job syncs.
c. job_finished_total: tracks the number of finished job.

2. add limitations pods create/delete operations per job sync, so that:
a. big job will not occupy controller too much time which may get other jobs stuck, due to too many pods create/delete at once.
b. gives much better predictability of metrics job_sync_duration_seconds

3. change the default feature gate value of "IndexedJob" from false to true, to make it enabled by default in beta.

#### Which issue(s) this PR fixes:

fixes #98434 

#### Special notes for your reviewer:
This is a requirement to graduate indexed job to Beta
The PR of changing hostname and podname is https://github.com/kubernetes/kubernetes/pull/101601

#### Does this PR introduce a user-facing change?

```release-note
Add three metrics to job controller to monitor if Job works in a healthy condition.
IndexedJob promoted to Beta
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2214-indexed-job
```

### CODE REVIEW

**Comment by AliceZhang2016** [comment]
cc @alculquicondor

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:419]: I think we can do this inside `syncHandler` instead. That way, we don't have to query the joblister twice.
You can do it in a `defer` function.

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [inline pkg/controller/job/metrics/metrics.go:40]: We should have result here too. We do this in scheduler as well https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/metrics/metrics.go#L51-L57

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [inline pkg/controller/job/metrics/metrics.go:43]: It doesn't matter if a job syncs several times, as the job key is not a label.

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [inline pkg/controller/job/metrics/metrics.go:23]: The name of this package suggests it's deprecated. We can probably use prometheus as suggested here https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md#quick-start

Comment (AliceZhang2016) [reply pkg/controller/job/metrics/metrics.go:23]: The code of package metrics seems only wraps the prometheus, it also uses prometheus inside. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/histogram.go#L95

File: pkg/controller/job/metrics/metrics.go
Comment (AliceZhang2016) [reply pkg/controller/job/metrics/metrics.go:43]: I'll remove this comment and keep NewCounterVec

File: pkg/controller/job/job_controller.go
Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:419]: The `JobSyncNum` also needs the job `mode` information. If we don't want to query the joblister again outside the syncHandler, we need to add several lines to increase `JobSyncNum` in `syncHandler` where it returns an error or nil. 
I think it maybe more concise and clear to keep the current code, unless query the joblister costs a lot.

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:419]: JobSyncNum and JobSyncLatency should be updated in the same scenarios: when the job was not deleted.

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [reply pkg/controller/job/metrics/metrics.go:23]: In that sense, there is no need for using legacyregistry, just use prometheus directly. Note that the guide was updated in February, so it's pretty up to date.

Comment (AliceZhang2016) [reply pkg/controller/job/metrics/metrics.go:23]: but the example in that guide was updated 6 years ago.
https://github.com/kubernetes/kubernetes/blob/cd3299307d44665564e1a5c77d0daa0286603ff5/pkg/apiserver/apiserver.go#L53

File: pkg/controller/job/job_controller.go
Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:419]: yep. and they both need to know if the syncJob returns an error or not. Use `defer` function can not get error information

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go]: yes, it can :)

You have to name the return values

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [reply pkg/controller/job/metrics/metrics.go:23]: @logicalhan what is the up-to-date recommendation?

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:446]: nit: name them `forget` and `err`. If you are worried about name conflicts, perhaps you can use `rErr`. I think `forget` should be fine to reuse within the function.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:491]: perhaps we can have this variable outside of the defer so that it can be shared with the rest of the metrics below.

File: pkg/controller/job/metrics/metrics.go
Comment (logicalhan) [reply pkg/controller/job/metrics/metrics.go:23]: You can't use prometheus directly, we actually have CI checks which should block you from merging PRs that use prometheus. The legacyregistry is deprecated because we want to eventually move off of using globals, which should happen hopefully this upcoming release.

Comment (logicalhan) [reply pkg/controller/job/metrics/metrics.go:23]: So the recommendation is to use the wrappers (from component base) and currently legacyregistry unless you are exposing a new metrics endpoint. I've submitted a PR to update the docs (https://github.com/kubernetes/community/pull/5737), sorry for the confusion.

**Review by logicalhan — COMMENTED** [review]
/lgtm
(from sig instrumentation) sorry about the confusion in the docs.

File: pkg/controller/job/metrics/metrics.go
Comment (logicalhan) [inline pkg/controller/job/metrics/metrics.go:40]: what's the cardinality of mode?

Comment (AliceZhang2016) [reply pkg/controller/job/metrics/metrics.go:40]: mode can be one of {Indexed, NonIndexed}, which indicates if it is an indexed job. I can add comment on label choice here.

Comment (logicalhan) [reply pkg/controller/job/metrics/metrics.go:40]: that'd be great, thanks!

**Comment by alculquicondor** [comment]
/hold
for KEP approval kubernetes/enhancements#2616

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:486]: rename it to `mode` or `completionMode`

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:809]: better crop the array podsToDelete.

Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:809]: but modify rmAtLeast before entering activePodForRemoval may reduce calculate operations in that function, why better crop podsToDelete?

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:809]: because activePodsForRemoval might return more than `rmAtLeast`.

Comment (AliceZhang2016) [reply pkg/controller/job/job_controller.go:809]: ah yes, you're right. Thanks for pointing out.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:808]: this can be simplified a bit:

```golang
if len(podsToDelete) > MaxPodCreateDeletePerSync {
  podsToDelete = podsToDelete[:MaxPodCreateDeletePerSync]
}
```

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:65]: we don't need this exported (so, lowercase is enough).

**Review by alculquicondor — COMMENTED** [review]
/hold cancel
KEP is approved

/lgtm

**Review by adtac — COMMENTED** [review]
/lgtm

File: pkg/apis/batch/types.go
Comment (ahg-g) [inline pkg/apis/batch/types.go:187]: did we actually make this change?

Comment (AliceZhang2016) [reply pkg/apis/batch/types.go:187]: https://github.com/kubernetes/kubernetes/pull/101601

Comment (ahg-g) [reply pkg/apis/batch/types.go:187]: please mention in the notes to reviewers that the change is being made in a separate PR

Comment (AliceZhang2016) [reply pkg/apis/batch/types.go:187]: done

**Comment by ahg-g** [comment]
can you please squash

**Comment by alculquicondor** [comment]
the commits contain 3 logical splits. I prefer not to squash.

**Review by janetkuo — APPROVED** [review]

**Review by erictune — APPROVED** [review]

**Comment by alculquicondor** [comment]
any reason why pkg/apis is not owned by api-approvers @liggitt? 😓

**Comment by liggitt** [comment]
> any reason why pkg/apis is not owned by api-approvers @liggitt? 😓

no reason, looks like I missed it. a PR to fix the package would be welcome

**Comment by AliceZhang2016** [comment]
@liggitt Then currently before the pkg/apis owned by api-approvers, could you please take a look of this PR to see if it can be approved. I find that you are one of the approvers in pkg/apis/OWNERS. Thanks

File: pkg/apis/batch/types.go
Comment (liggitt) [inline pkg/apis/batch/types.go:188]: is this accomplished by setting the pod.spec.hostname field? is that valid for all combinations of $jobname and $index (e.g. the longest possible job name and a multi-digit index)?

Comment (ahg-g) [reply pkg/apis/batch/types.go:188]: This is similar to statefulset, hostname is set to the pod name which is

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_utils.go#L88

Comment (ahg-g) [reply pkg/apis/batch/types.go:188]: It seems that we don't validate that the job name is a valid DNS name like we do for statefulset: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/apis/apps/validation/validation.go#L45

Comment (alculquicondor) [reply pkg/apis/batch/types.go:188]: It is through the hostname field.

It doesn't look like we validate the length of the StatefulSet name to account for every possible index.
Should we do it just if the Job.spec.completionMode=Indexed (alpha in 1.21).

Comment (ahg-g) [reply pkg/apis/batch/types.go:188]: Yes, I believe we can do this for the indexed completion mode only since it is still alpha, and by that we don't break job in non-indexed mode.

Comment (ahg-g) [reply pkg/apis/batch/types.go:188]: lets continue the discussion in the other PR

File: pkg/apis/batch/validation/validation.go
Comment (alculquicondor) [inline pkg/apis/batch/validation/validation.go:93]: hostname

File: pkg/apis/batch/validation/validation.go
Comment (alculquicondor) [inline pkg/apis/batch/validation/validation.go:91]: we add one more dash to the `generateName`: `$(job-name)-$(index)-`. So maybe 12 is a better number.
However, I'm not sure what happens if `generateName` already has 63 characters. @liggitt ?

File: pkg/apis/batch/validation/validation.go
Comment (AliceZhang2016) [reply pkg/apis/batch/validation/validation.go]: Done.

File: pkg/apis/batch/validation/validation.go
Comment (alculquicondor) [reply pkg/apis/batch/validation/validation.go:91]: It looks like apiserver handles the truncation of generateName to make space for the random characters.

/lgtm

Leaving the unhold to Jordan.

File: pkg/apis/batch/validation/validation.go
Comment (alculquicondor) [inline pkg/apis/batch/validation/validation.go:94]: Pod hostname length limit

Comment (AliceZhang2016) [reply pkg/apis/batch/validation/validation.go]: Done

File: pkg/apis/batch/types.go
Comment (ahg-g) [reply pkg/apis/batch/types.go:188]: We should also check that it is a valid DNs name, not just the length. @johnbelamaric can we use [`IsDNS1123Label`](https://github.com/kubernetes/kubernetes/blob/31030820be979ea0b2c39e08eb18fddd71f353ed/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L187)?

Comment (liggitt) [reply pkg/apis/batch/types.go:188]: we should make sure that whatever construction we make from the name and biggest possible index will pass the same validation as the hostname field

I think that means 
`ValidateDNS1123Label(name + "-" + biggest possible index)`

also, if the biggest possible index can change, then that validation should be done on updates to indexed jobs that attempt to change the biggest possible index

Comment (alculquicondor) [reply pkg/apis/batch/types.go:188]: @ahg-g the name is already validated. The check is the same for every controller (including statefulset).

So you are proposing we do the validation dynamic, based on the `Job.spec.completions`. That works for me. It will give more room for Job name, as not all will use the max possible int32.

Today, the number of completions is immutable, but I'm thinking of changing that in the next release. So, let's check during Update.

Comment (alculquicondor) [reply pkg/apis/batch/types.go:188]: New validation LGTM. Note that the same validation function is also called during Update, so no need for any extra changes in the Update strategy.

File: pkg/apis/batch/validation/validation.go
Comment (liggitt) [inline pkg/apis/batch/validation/validation.go:96]: completions can be 0:

https://github.com/kubernetes/kubernetes/blob/a5cf298a95b4a60789a6ba1caab931bdc980aa2d/pkg/apis/batch/validation/validation.go#L120-L122

https://github.com/kubernetes/kubernetes/blob/a5cf298a95b4a60789a6ba1caab931bdc980aa2d/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L81-L88

so completions - 1 would be -1?

Comment (AliceZhang2016) [reply pkg/apis/batch/validation/validation.go:96]: When the completions=0, there will be no pods created, thus no pod will have a hostname with index=-1 as suffix. 

But yes, there is one case with different result after judging `completions>0`. When the job name > 60 characters, it will cause error here as the length exceeds 63 after appending `--1`.

Maybe better to add restrict completions to be positive? (referred to https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs, there is a sentence “specify a **non-zero positive** value for .spec.completions.”) Besides, a job with `completions=0` seems useless.

Comment (alculquicondor) [reply pkg/apis/batch/validation/validation.go:96]: Since no pods will be created, you can just skip the check for completions = 0. But yes, I agree that completions=0 doesn't make sense. But that's not something we can change due to backwards compatibility.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/apis/batch/validation/validation.go
Comment (liggitt) [reply pkg/apis/batch/validation/validation.go:96]: > When the completions=0, there will be no pods created, thus no pod will have a hostname with index=-1 as suffix.

then I'd either skip this check for completions=0 or set the minimum index we check length for to 0

Comment (AliceZhang2016) [reply pkg/apis/batch/validation/validation.go:96]: In the latest commit, this check is skipped when completions=0

**Comment by liggitt** [comment]
validation change looks good to me. eric's approve is now effective

**Pull request merged**
